lib/src/client/impl/exchange_impl.dart

```diff
@@ -22,7 +22,9 @@ class _ExchangeImpl implements Exchange {
   }
 
   void publish(Object message, String routingKey, { MessageProperties properties, bool mandatory : false, bool immediate : false}) {
-    if (type != ExchangeType.FANOUT && (routingKey == null || routingKey.isEmpty)) {
+    if(type is CustomExchangeType && (type as CustomExchangeType).routingKeyRequired && (routingKey == null || routingKey.isEmpty)){
```

Comment:
How about moving the routingKeyRequired property into the Exchange or BaseExchange class and have the FANOUT constant set its value to true? You could then simplify L25:L27 to a single if check